### DISCUSSION TITLE
`cuda.core` v0.1.0

### DISCUSSION THREAD

**Discussion by leofang** [NVIDIA/cuda-python#232] [Announcements]
We are pleased to introduce a new subpackage `cuda.core` to Python users, the first very step toward "pythonic CUDA"! 🚀 

This is the initial beta release, with more beta releases planned to follow.

## Mission
The new module `cuda.core` is aimed to
- provide idiomatic ("pythonic") access to CUDA Driver, Runtime, and JIT compiler toolchain
- focus on developer productivity by ensuring end-to-end CUDA development can be performed quickly and entirely in Python
- avoid homegrown Python abstractions for CUDA for new Python GPU libraries starting from scratch
- ease developer burden of maintaining and catching up with latest CUDA features
- flatten the learning curve for current and future generations of CUDA developers

## Development model
`cuda.core` adopts an iterative development process to gradually cover major CUDA features in a pythonic fashion, while collecting/incorporating early adopters' feedback for the design. The `cuda.core` design can and may change across beta releases, and will be declared stable when we reach GA (General Access, v1.0.0).

## Key Features and Enhancements
- Idiomatic/pythonic access to CUDA Driver, objects covered
   - `Device`
   - `Stream`
      - Also introduce the `__cuda_stream__` protocol
   - `Event`
   - `Buffer`
   - `launch()` & `LaunchConfig`
      - support Python scalars and NumPy/ctypes fixed-width scalars as kernel arguments
   - `Program`
   - `ObjectCode`
   - `Kernel`

Code samples are available on GitHub: https://github.com/NVIDIA/cuda-python/tree/main/cuda_core/examples

All functionalities are currently hosted under the `cuda.core.experimental` namespace. Once the features become stable they will be moved out of `experimental`.

## Release notes
See the release notes (and full documentation) at https://nvidia.github.io/cuda-python/cuda-core/0.1.0/.

<hr /><em>This discussion was created from the release <a href='https://github.com/NVIDIA/cuda-python/releases/tag/cuda-core-v0.1.0'>cuda.core v0.1.0</a>.</em>

**Comment by jianruyin** [comment]
Congratulations on your release. Could you please introduce more about the necessary whole stack of CUDA and where is cuda-python, and where will it go? Thanks.

**Reply by leofang** [reply]
Yes, thanks @jianruyin for asking. I've added a bit more detailed post to the OP above.

Regarding the question 
> and where will it go? 

as noted in the updated OP, we plan to gradually cover major CUDA features in a pythonic fashion, but we would love to hear about what users need too. Our plan/roadmap is tracked in [various milestones](https://github.com/NVIDIA/cuda-python/milestones).
